File: C Language/Remove_Kth_Node_From_End_LinkedList.c
//Remove Duplicates from Sorted Singly Linked List

#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node * next;
    struct node * prev;
};

struct node *head = NULL;

void linkedListTraversal(struct node* ptr)
{
    ptr = head;
    while(ptr != NULL)
    {
        printf("%d ", ptr -> data);
        ptr = ptr -> next;
    }
}

void createList(int data)
{
    struct node *ptr = (struct node*)malloc(sizeof(struct node));
    ptr -> data = data;
    ptr -> next = head;	
    head = ptr;
}

struct node* removeNthNode(struct node* ptr, int key)
{
    //find the nth node from end
    ptr = head;
    int n = 0;
    if(ptr == NULL)
    {
        return;
    }
    while(ptr)
    {
        ptr = ptr -> next;
        n++;
    }
    if(n >= key)
    {
        ptr = head;
        for(int i= 0; i<n-key; i++)
        {
            ptr = ptr -> next;
        }
    }
    
    //remove the nth node from end
    struct node* p = head;
    struct node* q = head -> next;
    while((q -> data != ptr -> data) && (q -> next != NULL))
    {
        p = p -> next;
        q = q -> next;
    }
    if(q -> data == ptr -> data)
    {
        p -> next = q -> next;
        free(q);
    }
    return head;
}


int main()
{
   
    createList(92);
    createList(70);
    createList(70);
    createList(36);
    createList(28);
    createList(28);
    createList(12);
    printf("Linked List before removal\n");
    linkedListTraversal(head);
    
    printf("\n\nLinked List after removal\n");
    head = removeNthNode(head, 4);
    linkedListTraversal(head);
    return 0;
}

File: Python/Remove_Kth_Node_From_End_LinkedList.py
class Node:
	def __init__(self, data):
		self.data = data
		self.next = None
		self.prev = None

class LinkedList:
	def __init__(self):
		self.head = None
			
	def push(self, data):
		newNode = Node(data)
		newNode.next = self.head
		self.head = newNode
		
	def printList(self, ptr):
		ptr = self.head
		if ptr is None:
			print('List has no element')
			return
		else:
			temp = ptr
			while(temp):
				print(temp.data, end=' ')
				temp = temp.next
				
	def removeNthNode(self, key):
		#find the nth node from end
		ptr = self.head
		n = 0
		if ptr is None:
			return
		while(ptr):
		      ptr = ptr.next
		      n += 1
		if n >= key:
			ptr = self.head
			for i in range(0, n-key):
				ptr = ptr.next
		
		#remove the nth node from end
		p = self.head
		q = self.head.next
		while q.data != ptr.data and q.next != None:
				p = p.next
				q = q.next
		if q.data == ptr.data:
			p.next = q.next
		if self.head.data == ptr.data:
						self.head = self.head.next
		return
					
llist = LinkedList()

llist.push(92)
llist.push(70)
llist.push(36)
llist.push(28)
llist.push(12)

print("Linked List before removal")
llist.printList(llist.head)

print("\n\nLinked List after removal")
result = llist.removeNthNode(3)
llist.printList(result)